packages/console/src/components/UserRolesTransfer/components/SourceRolesBox/index.tsx

```diff
@@ -0,0 +1,117 @@
+import type { RoleResponse } from '@logto/schemas';
+import type { ChangeEvent } from 'react';
+import { useEffect, useRef, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import useSWR from 'swr';
+
+import Search from '@/assets/images/search.svg';
+import Pagination from '@/components/Pagination';
+import TextInput from '@/components/TextInput';
+import * as transferLayout from '@/scss/transfer.module.scss';
+import { buildUrl } from '@/utilities/url';
+
+import SourceRoleItem from '../SourceRoleItem';
+import * as styles from './index.module.scss';
+
+type Props = {
+  userId: string;
+  selectedRoles: RoleResponse[];
+  onChange: (value: RoleResponse[]) => void;
+};
+
+const pageSize = 20;
+const searchDelay = 500;
+
+const SourceRolesBox = ({ userId, selectedRoles, onChange }: Props) => {
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+
+  const [pageIndex, setPageIndex] = useState(1);
+  const [keyword, setKeyword] = useState('');
+
+  const timerRef = useRef<NodeJS.Timeout>();
+
+  const clearTimer = () => {
+    if (timerRef.current) {
+      clearTimeout(timerRef.current);
+    }
+  };
+
+  useEffect(() => {
+    return () => {
+      clearTimer();
+    };
+  }, []);
+
+  const url = buildUrl('/api/roles', {
+    excludeUserId: userId,
+    page: String(pageIndex),
+    page_size: String(pageSize),
+    ...conditional(keyword && { search: `%${keyword}%` }),
+  });
+
+  const { data } = useSWR<[RoleResponse[], number]>(url);
+
+  const [dataSource = [], totalCount] = data ?? [];
+
+  const isRoleSelected = (role: RoleResponse) =>
+    selectedRoles.findIndex(({ id }) => role.id === id) >= 0;
+
+  const handleSearchInput = (event: ChangeEvent<HTMLInputElement>) => {
+    clearTimer();
+    // eslint-disable-next-line @silverhand/fp/no-mutation
+    timerRef.current = setTimeout(() => {
+      setPageIndex(1);
+      setKeyword(event.target.value);
+    }, searchDelay);
```

Comment:
this is a common pattern of debounce or something too, a hook?

Reply:
Good call, updated.